File: cpp/binarySearch/time_based_store.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long
#define pb push_back
#define all(x) x.begin(), x.end()
#define endl '\n'

class TimeMap {
  public:
    unordered_map<string, vector<pair<int, string>>> mymap;
    TimeMap() {
    }

    void set(string key, string value, int timestamp) {
        mymap[key].push_back({ timestamp, value });
    }

    string get(string key, int timestamp) {
        int left = 0;
        int right = mymap[key].size() - 1;
        vector<pair<int, string>>& cp = mymap[key];
        if (cp.size() == 0) return "";
        while (left <= right) {
            int mid = (right + left) / 2;
            if (timestamp >= cp[mid].first)
                left = mid + 1;
            else
                right = mid - 1;
        }

        if (right > 0) {
            return cp[right].second;
        }
        return "";
    }
};

int main() {

    ios::sync_with_stdio(0);
    cin.tie(0);

    return 0;
}

File: python/binarySearch/timeBasedStore.py
import sys
import threading
from typing import List, Tuple, Dict, Set
from collections import defaultdict, deque


class TimeMap:

    def __init__(self):
        self.mymap = {}

    def set(self, key, value, timestamp):
        self.mymap.setdefault(key, []).append((value, timestamp))

    def get(self, key, timestamp):
        if key not in self.mymap:
            return ""

        arr = self.mymap[key]
        left = 0
        right = len(arr) - 1

        while left <= right:
            mid = (right + left) // 2
            if timestamp >= arr[mid][1]:
                left = mid + 1
            else:
                right = mid - 1

        if right < 0:
            return ""

        return arr[right][0]


def main():

    timeMap = TimeMap()
    timeMap.set("alice", "happy", 1)
    print(timeMap.get("alice", 1))
    print(timeMap.get("alice", 2))
    timeMap.set("alice", "sad", 3)
    print(timeMap.get("alice", 3))


if __name__ == "__main__":
    threading.Thread(target=main).start()
